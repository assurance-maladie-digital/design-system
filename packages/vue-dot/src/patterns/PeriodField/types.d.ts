export interface PeriodValue {
	from: string | null;
	to: string | null;
}
