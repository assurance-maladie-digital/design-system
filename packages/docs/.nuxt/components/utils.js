// nuxt/nuxt.js#8607
export function wrapFunctional(options) {
  if (!options || !options.functional) {
    return options
  }

  const propKeys = Array.isArray(options.props) ? options.props : Object.keys(options.props || {})

  return {
    render(h) {
      const attrs = {}
      const props = {}

      for (const key in this.$attrs) {
        if (propKeys.includes(key)) {
          props[key] = this.$attrs[key]
        } else {
          attrs[key] = this.$attrs[key]
        }
      }

      return h(options, {
        on: this.$listeners,
        attrs,
        props,
        scopedSlots: this.$scopedSlots,
      }, this.$slots.default)
    }
  }
}
