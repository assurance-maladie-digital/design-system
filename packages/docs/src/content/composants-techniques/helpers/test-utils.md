---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
