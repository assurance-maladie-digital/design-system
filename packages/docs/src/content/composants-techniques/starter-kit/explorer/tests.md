Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
