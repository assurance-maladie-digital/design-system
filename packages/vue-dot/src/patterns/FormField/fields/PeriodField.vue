<template>
	<VLayout
		wrap
		row
		class="ma-n3"
	>
		<VCol
			cols="12"
			md="6"
		>
			<DatePicker
				v-model="periodValue.from"
				v-bind="field.metadata"
				:vuetify-options="field.metadata.from"
				@change="changeDate"
			/>
		</VCol>

		<VCol
			cols="12"
			md="6"
		>
			<DatePicker
				v-model="periodValue.to"
				v-bind="field.metadata"
				:vuetify-options="metadataTo"
				:start-date="periodValue.from"
				@change="changeDate"
			/>
		</VCol>
	</VLayout>
</template>

<script lang="ts">
	import Vue from 'vue';
	import Component from 'vue-class-component';

	import DatePicker from '../../DatePicker';

	import FieldComponent from '../mixins/fieldComponent';

	import { PeriodValue } from '../types';

	import cloneDeep from 'lodash.clonedeep';

	/** Form period field to enter a period */
	@Component<PeriodField>({
		components: {
			DatePicker
		},
		watch: {
			// Listen the current field value for the component
			'field.value': {
				handler(value: PeriodValue | null) {
					if (value) {
						this.periodValue = value;
					}
				},
				immediate: true,
				deep: true
			}
		}
	})
	export default class PeriodField extends FieldComponent {
		periodValue: PeriodValue = {
			from: null,
			to: null
		};

		get metadataTo() {
			const datePicker = {
				min: this.periodValue.from
			};

			return {
				datePicker,
				...this.field.metadata.to
			};
		}

		/** Emit the new value when started or ended date change */
		changeDate() {
			// Reset the end date if selected start date greater
			if (
				this.periodValue.from &&
				this.periodValue.to &&
				new Date(this.periodValue.from) > new Date(this.periodValue.to)
			) {
				this.periodValue.to = null;
			}

			// Set the new value of the field
			const field = cloneDeep(this.field);
			field.value = this.periodValue;

			this.$nextTick(() => {
				this.emitChangeEvent(field);
			});
		}
	}
</script>
