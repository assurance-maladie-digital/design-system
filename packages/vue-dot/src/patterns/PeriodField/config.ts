export const config = {
	from: {
		textField: {
			label: 'DÃ©but',
			hint: 'Format JJ/MM/AAAA'
		}
	},
	to: {
		textField: {
			label: 'Fin',
			hint: 'Format JJ/MM/AAAA'
		}
	}
};
