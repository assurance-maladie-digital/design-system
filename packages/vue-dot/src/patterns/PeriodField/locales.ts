export const locales = {
	greaterThanTo: 'La date saisie doit être supérieure à la date de début.'
};
