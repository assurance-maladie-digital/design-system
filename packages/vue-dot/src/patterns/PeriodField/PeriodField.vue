<template>
	<div class="vd-period-field d-flex flex-wrap max-width-none ma-n2">
		<DatePicker
			v-model="periodValue.from"
			:vuetify-options="datePickerOptions.from"
			:outlined="outlined"
			:disabled="disabled"
			text-field-class="vd-period-field-picker flex-grow-1 ma-2"
			@change="dateUpdated"
		/>

		<DatePicker
			v-model="periodValue.to"
			:outlined="outlined"
			:disabled="disabled"
			:start-date="periodValue.from"
			:vuetify-options="toPickerOptions"
			text-field-class="vd-period-field-picker flex-grow-1 ma-2"
			@change="dateUpdated"
		/>
	</div>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { customizable, Options } from '../../mixins/customizable';

	import { PeriodValue } from './types';
	import { config } from './config';
	import { locales } from './locales';

	import deepMerge from 'deepmerge';

	const Props = Vue.extend({
		props: {
			value: {
				type: Object as PropType<PeriodValue>,
				default: () => ({
					from: null,
					to: null
				})
			},
			outlined: {
				type: Boolean,
				default: false
			},
			disabled: {
				type: Boolean,
				default: false
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config));

	@Component<PeriodField>({
		model: {
			prop: 'value',
			event: 'change'
		},
		watch: {
			value: {
				handler(value: PeriodValue): void {
					if (value) {
						this.periodValue = value;

						return;
					}

					this.periodValue = {
						from: null,
						to: null
					};

				},
				immediate: true,
				deep: true
			}
		}
	})
	export default class PeriodField extends MixinsDeclaration {
		periodValue = {} as PeriodValue;

		errorMessages: string[] = [];

		get datePickerOptions(): Options {
			return deepMerge<Options>(config, this.options);
		}

		get toPickerOptions(): Options {
			const { to } = this.datePickerOptions;

			to.textField = {
				...to.textField,
				errorMessages: this.errorMessages
			};

			return to;
		}

		async dateUpdated(): Promise<void> {
			const fromGreaterThanTo = (
				this.periodValue.from &&
				this.periodValue.to &&
				new Date(this.periodValue.from) > new Date(this.periodValue.to)
			);

			if (fromGreaterThanTo) {
				this.errorMessages = [locales.greaterThanTo];
			} else {
				this.errorMessages = [];
			}

			if (!this.periodValue.from || !this.periodValue.to) {
				return;
			}

			await this.$nextTick();

			this.emitChangeEvent();
		}

		emitChangeEvent(): void {
			this.$emit('change', this.periodValue);
		}
	}
</script>

<style lang="scss" scoped>
	@import '@cnamts/design-tokens/dist/tokens';

	.vd-period-field :deep(.vd-period-field-picker) {
		max-width: $vd-input-medium;
	}
</style>
