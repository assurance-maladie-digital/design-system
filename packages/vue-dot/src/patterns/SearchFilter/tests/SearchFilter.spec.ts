import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';
import { html } from '@/tests/utils/html';

import PeriodField from '../';

let wrapper: Wrapper<Vue>;

describe('PeriodField', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(PeriodField);

		expect(html(wrapper)).toMatchSnapshot();
	});
});
