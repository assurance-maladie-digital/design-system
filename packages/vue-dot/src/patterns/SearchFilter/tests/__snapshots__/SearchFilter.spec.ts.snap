// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeriodField renders correctly 1`] = `
<div>
  <vtextfield-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" backgroundcolor="" label="Rechercher" loaderheight="2" clearable="true" clearicon="$clear" outlined="true" type="text"></vtextfield-stub>



  <!---->
</div>
`;
